### PR TITLE
Standardize approach to dependencies

### PR DESCRIPTION
**Pull request by potiuk**
Approach to dependencies we had (especially with regards to
upper bounds) was pretty random so far. This PR attempts to
describe the rules discussed in the devlist discussion - including
review and update of all dependencies to match the policies.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -292,65 +302,72 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
...
 ]
 google = [
+    # Google has very clear rules on what dependencies should be used. All the limits below
+    # follow strict guidelines of Google Libraries as quoted here:
+    # While this issue is open, dependents of google-api-core, google-cloud-core. and google-auth
+    # should preserve >1, <3 pins on these packages.
+    # https://github.com/googleapis/google-cloud-python/issues/10566
+    # Some of Google Packages are limited to <2.0.0 because 2.0.0 releases of the libraries
+    # Introduced breaking changes across the board. Those libraries should be upgraded soon
+    # TODO: Upgrade all Google libraries that are limited to <2.0.0
```
Comment (ashb) [inline setup.py:312]: Given v2.0 if the libraries _did_ introduce breaking changes across the board (or so I understand?), I wonder if this is a case where w _should_ keep the `<2` restriction?

Comment (potiuk) [reply setup.py:312]: Yep . I think I kept all `< 2.0.0` where they were there and only removed the <3, <4 which are future-unknown-if-ever-happens-and-what-the-scope-of-breakage-will-be ones. Let me re-review if I did it right.

Comment (potiuk) [reply setup.py:312]: Yep. All the `<2.0.0` were left. I only removed <3 and <4.

This was actually even announced and "warned" before by Google that 2.0.0 for all the libraries will be breaking (they changed the way how 2.0.0 libraries were generated (they are mostly automatically generated). So I do not *really* expect similar kind of breakage even in the future versions (and we do not even know if they will ever be 3s or 4s for those)

**Comment by potiuk** [comment]
I think that one is pretty read if we will be ok with introducing the policies as I proposed .

File: README.md [hunk 2]
```diff
@@ -308,6 +309,54 @@ They are based on the official release schedule of Python and Kubernetes, nicely
...
+
+Whenever we upper-bound such a dependency, we should always comment why we are doing it - i.e. we should have
+a good reason why dependency is upper-bound. And we should also mention what is the condition to remove the
+binding.
+
+### Approach for dependencies for Airflow Core
+
+Those `extras` and `providers` dependencies are maintained in `setup.cfg`.
+
+There are few dependencies that we decided are important enough to upper-bind them by default, as they are
```
Comment (ashb) [inline README.md]:
```suggestion
There are few dependencies that we decided are important enough to upper-bound them by default, as they are
```

**Review by ashb — APPROVED** [review]
Lets see how this shakes out.

**Comment by potiuk** [comment]
Well. We do not really risk much. We can always limit troublesome deps if we find them troublesome. 

We always have "escape hatch" -> use constraints. 

Plus more and more people use images actually wich makes them naturally constrained to what we produce. 

And we did not really had many of those in the first place actually.

**Pull request merged**